src/cmd/tools/verify_data_files/main/main.go

```diff
@@ -0,0 +1,473 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	golog "log"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+	"unicode/utf8"
+
+	"github.com/m3db/m3/src/cmd/tools"
+	"github.com/m3db/m3/src/dbnode/digest"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/storage/index/convert"
+	"github.com/m3db/m3/src/x/checked"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/pool"
+
+	"github.com/pborman/getopt"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const snapshotType = "snapshot"
+const flushType = "flush"
+
+func main() {
+	var (
+		optPathPrefix     = getopt.StringLong("path-prefix", 'p', "/var/lib/m3db", "Path prefix [e.g. /var/lib/m3db]")
+		optFailFast       = getopt.BoolLong("fail-fast", 'f', "Fail fast will bail on first failure")
+		optFixDir         = getopt.StringLong("fix-path-prefix", 'o', "/tmp/m3db", "Fix output path file prefix for fixed files [e.g. /tmp/m3db]")
+		optFixInvalidIDs  = getopt.BoolLong("fix-invalid-ids", 'i', "Fix invalid IDs will remove entries with IDs that have non-UTF8 chars")
+		optFixInvalidTags = getopt.BoolLong("fix-invalid-tags", 't', "Fix invalid tags will remove tags with name/values non-UTF8 chars")
+		optDebugLog       = getopt.BoolLong("debug", 'd', "Enable debug log level")
+	)
+	getopt.Parse()
+
+	logLevel := zapcore.InfoLevel
+	if *optDebugLog {
+		logLevel = zapcore.DebugLevel
+	}
+
+	logConfig := zap.NewDevelopmentConfig()
+	logConfig.Level = zap.NewAtomicLevelAt(logLevel)
+	log, err := logConfig.Build()
+	if err != nil {
+		golog.Fatalf("unable to create logger: %+v", err)
+	}
+
+	if *optPathPrefix == "" {
+		getopt.Usage()
+		os.Exit(1)
+	}
+
+	run(runOptions{
+		filePathPrefix: *optPathPrefix,
+		failFast:       *optFailFast,
+		fixDir:         *optFixDir,
+		fixInvalidIDs:  *optFixInvalidIDs,
+		fixInvalidTags: *optFixInvalidTags,
+		log:            log,
+	})
+}
+
+type runOptions struct {
+	filePathPrefix string
+	failFast       bool
+	fixDir         string
+	fixInvalidIDs  bool
+	fixInvalidTags bool
+	log            *zap.Logger
+}
+
+func run(opts runOptions) {
+	filePathPrefix := opts.filePathPrefix
+	log := opts.log
+
+	log.Info("creating bytes pool")
+	bytesPool := tools.NewCheckedBytesPool()
+	bytesPool.Init()
+
+	dataDirPath := fs.DataDirPath(filePathPrefix)
+
+	namespaces, err := dirFiles(dataDirPath)
+	if err != nil {
+		log.Fatal("could not read namespaces", zap.Error(err))
+	}
+
+	// Get all fileset files.
+	log.Info("discovering file sets",
```

Comment:
meganit: could build up a `fields := []zapcore.Fields{zap.Strings("namespaces", namespaces)}` and append to it as we go?